routing/router.go

```diff
@@ -1138,60 +1145,85 @@ func (r *ChannelRouter) networkHandler() {
 
 			// We'll ensure that any new blocks received attach
 			// directly to the end of our main chain. If not, then
-			// we've somehow missed some blocks. We don't process
-			// this block as otherwise, we may miss on-chain
-			// events.
+			// we've somehow missed some blocks. Here we'll catch
+			// up the chain with the latest blocks.
 			currentHeight := atomic.LoadUint32(&r.bestHeight)
-			if chainUpdate.Height != currentHeight+1 {
+			if chainUpdate.Height == currentHeight+1 {
+				err := r.updateGraphWithClosedChannels(
+					chainUpdate, chainUpdate.Height,
+				)
+				if err != nil {
+					log.Errorf("unable to prune graph "+
+						"with closed channels: %v", err)
+				}
+			} else {
 				log.Errorf("out of order block: expecting "+
-					"height=%v, got height=%v", currentHeight+1,
-					chainUpdate.Height)
-				continue
-			}
+					"height=%v, got height=%v",
+					currentHeight+1, chainUpdate.Height)
+
+				if chainUpdate.Height > currentHeight+1 {
+					outdatedHash, err := r.cfg.Chain.GetBlockHash(
+						int64(currentHeight),
+					)
+					if err != nil {
+						log.Errorf("unable to retrieve" +
+							" outdated block hash")
+					}
 
-			// Once a new block arrives, we update our running
-			// track of the height of the chain tip.
-			blockHeight := uint32(chainUpdate.Height)
-			atomic.StoreUint32(&r.bestHeight, blockHeight)
-			log.Infof("Pruning channel graph using block %v (height=%v)",
-				chainUpdate.Hash, blockHeight)
-
-			// We're only interested in all prior outputs that have
-			// been spent in the block, so collate all the
-			// referenced previous outpoints within each tx and
-			// input.
-			var spentOutputs []*wire.OutPoint
-			for _, tx := range chainUpdate.Transactions {
-				for _, txIn := range tx.TxIn {
-					spentOutputs = append(spentOutputs,
-						&txIn.PreviousOutPoint)
-				}
-			}
+					outdatedBlock := &chainntnfs.BlockEpoch{
+						Height: int32(currentHeight),
+						Hash:   outdatedHash,
+					}
 
-			// With the spent outputs gathered, attempt to prune
-			// the channel graph, also passing in the hash+height
-			// of the block being pruned so the prune tip can be
-			// updated.
-			chansClosed, err := r.cfg.Graph.PruneGraph(spentOutputs,
-				&chainUpdate.Hash, chainUpdate.Height)
-			if err != nil {
-				log.Errorf("unable to prune routing table: %v", err)
-				continue
-			}
+					epochClient := r.cfg.BlockNotifier
 
-			log.Infof("Block %v (height=%v) closed %v channels",
-				chainUpdate.Hash, blockHeight, len(chansClosed))
+					// The block notifier client should be
+					// nil unless we're running a unit test.
+					if epochClient == nil {
```

Comment:
Remove this and just `RegisterBlockEpochNtfn` here, unit tests can just use an appropriately mocked notifier.